mmaction/utils/gradcam_utils.py

```diff
@@ -0,0 +1,230 @@
+        Returns:
+            tuple[torch.Tensor, torch.Tensor]: (localization_map, preds)
+                localization_map (torch.Tensor): the localization map for
+                    input imgs.
+                preds (torch.Tensor): Model predictions for `inputs` with
+                    shape (batch_size, num_classes).
+        """
+        inputs['imgs'] = inputs['imgs'].clone()
+
+        # model forward & backward
+        preds = self.model(gradcam=True, **inputs)
+        if use_labels:
+            labels = inputs['label']
+            if labels.ndim == 1:
+                labels = labels.unsqueeze(-1)
+            score = torch.gather(preds, dim=1, index=labels)
+        else:
+            score = torch.max(preds, dim=-1)[0]
+        self.model.zero_grad()
+        score = torch.sum(score)
+        score.backward()
+
+        if self.is_recognizer2d:
+            # [batch_size, num_segments, 3, H, W]
+            b, t, _, h, w = inputs['imgs'].size()
+        else:
+            # [batch_size, num_crops*num_clips, 3, clip_len, H, W]
+            b1, b2, _, t, h, w = inputs['imgs'].size()
+            b = b1 * b2
+
+        gradients = self.target_gradients
+        activations = self.target_activations
+        if self.is_recognizer2d:
+            # [B*Tg, C', H', W']
+            b_tg, c, _, _ = gradients.size()
+            tg = b_tg // b
+        else:
+            # source shape: [B, C', Tg, H', W']
+            _, c, tg, _, _ = gradients.size()
+            # target shape: [B, Tg, C', H', W']
+            gradients = gradients.permute(0, 2, 1, 3, 4)
+            activations = activations.permute(0, 2, 1, 3, 4)
+
+        # calculate & resize to [B, 1, T, H, W]
+        weights = torch.mean(gradients.view(b, tg, c, -1), dim=3)
+        weights = weights.view(b, tg, c, 1, 1)
+        activations = activations.view([b, tg, c] +
+                                       list(activations.size()[-2:]))
+        localization_map = torch.sum(
+            weights * activations, dim=2, keepdim=True)
+        localization_map = F.relu(localization_map)
+        localization_map = localization_map.permute(0, 2, 1, 3, 4)
+        localization_map = F.interpolate(
+            localization_map,
+            size=(t, h, w),
+            mode='trilinear',
+            align_corners=False)
+
+        # Normalize the localization map.
+        localization_map_min, localization_map_max = (
+            torch.min(localization_map.view(b, -1), dim=-1, keepdim=True)[0],
+            torch.max(localization_map.view(b, -1), dim=-1, keepdim=True)[0])
+        localization_map_min = torch.reshape(
+            localization_map_min, shape=(b, 1, 1, 1, 1))
+        localization_map_max = torch.reshape(
+            localization_map_max, shape=(b, 1, 1, 1, 1))
+        # delta must be small enough.
+        # make sure `localization_map_max - localization_map_min >> delta`
+        delta = 1e-20
```

Comment:
It is not good to use hardcoded value for `delta`. you can set it as keyword argument in the function.

```python
def _calculate_localization_map(self, inputs, use_labels, delta=1e-20):
```

And do not forget to update the docstring, thx : )